lib/engine/train.rb

```diff
@@ -55,6 +55,12 @@ def variant=(new_variant)
       @variant.each { |k, v| instance_variable_set("@#{k}", v) }
     end
 
+    def update_end_of_life(rusts_on, obsolete_on)
```

Comment:
why is this only called in 18mex? should this just be done inline in 18mex? will this be used for other games?

Reply:
@tobymao This is to change the configuration (when a train is obsoleted, rusted) so I do not know if that many games will use it.
But it is at least used by optional rules in other Mark Derrick games, at least 18AL. I have 18AL optional rules ready since before - just have to update them with using this.

If it were to be done in the game, don't I have to change the accessors (rusts_on, obsolete_on, variants)? The rusts_on and obsolete_on was changed to read-write access because of the 18MEX optional rule need so this is going back to not exposing it.  Also if doing it outside does expose details on how @variants is represented, which might be unnecessary.

But there is another potential here - it is WHY is the @variants created for a game that does not have any variant? The bug was caused by the train class creating variant duplication so maybe I should look into that instead? If I fix it so variant is not created then I can do as I did before, that is making rusts_on and obsolete_on accessible and update them unline in 18MEX. That is probably a better fix.

Reply:
Thinking about it it is probably just to avoid initiating @variants in case :variants is nil.

Reply:
sound sgood